File: packages/tonic-core/src/hydration/llmService.ts
import { HydrationRuntimeError } from "./errors";
import { resolveHydrationLlmConfig, type HydrationLlmConfig } from "./llmConfig";

export type HydrationLlmRole = "system" | "user";

export type HydrationLlmMessage = {
  role: HydrationLlmRole;
  content: string;
};

export interface HydrationLlmService {
  readonly providerId: string;
  readonly model: string;
  generateJson<T>(args: {
    messages: HydrationLlmMessage[];
    schemaName: string;
    schemaDescription: string;
  }): Promise<T>;
}

function extractJsonObject(text: string): string {
  const trimmed = text.trim();
  if (trimmed.startsWith("{") && trimmed.endsWith("}")) {
    return trimmed;
  }
  const start = trimmed.indexOf("{");
  const end = trimmed.lastIndexOf("}");
  if (start >= 0 && end > start) {
    return trimmed.slice(start, end + 1);
  }
  throw new HydrationRuntimeError("LLM response did not contain a JSON object.");
}

class OpenAiHydrationLlmService implements HydrationLlmService {
  readonly providerId = "openai";
  readonly model: string;
  readonly config: HydrationLlmConfig;

  constructor(config: HydrationLlmConfig) {
    if (!config.apiKey) {
      throw new HydrationRuntimeError(
        "TONIC_HYDRATION_LLM_MODE=openai requires TONIC_HYDRATION_LLM_API_KEY or OPENAI_API_KEY.",
      );
    }
    this.config = config;
    this.model = config.model;
  }

  async generateJson<T>(args: {
    messages: HydrationLlmMessage[];
    schemaName: string;
    schemaDescription: string;
  }): Promise<T> {
    const controller = new AbortController();
    const timeout = setTimeout(() => controller.abort(), this.config.timeoutMs);
    try {
      const response = await fetch(`${this.config.baseUrl}/chat/completions`, {
        method: "POST",
        headers: {
          "Content-Type": "application/json",
          Authorization: `Bearer ${this.config.apiKey}`,
        },
        body: JSON.stringify({
          model: this.config.model,
          temperature: 0,
          response_format: { type: "json_object" },
          messages: [
            ...args.messages,
            {
              role: "system",
              content:
                `Return only a JSON object for schema '${args.schemaName}'. `
                + `Required shape: ${args.schemaDescription}`,
            },
          ],
        }),
        signal: controller.signal,
      });
      const text = await response.text();
      if (!response.ok) {
        throw new HydrationRuntimeError(`OpenAI HTTP ${response.status}: ${text}`);
      }
      const payload = JSON.parse(text) as {
        choices?: Array<{ message?: { content?: string | Array<{ type?: string; text?: string }> } }>;
      };
      const content = payload.choices?.[0]?.message?.content;
      const output =
        typeof content === "string" ? content
        : Array.isArray(content) ? content.map((entry) => entry.text ?? "").join("")
        : "";
      return JSON.parse(extractJsonObject(output)) as T;
    } catch (error) {
      if (error instanceof HydrationRuntimeError) {
        throw error;
      }
      throw new HydrationRuntimeError(
        `Failed to generate structured hydration output via ${this.providerId}: ${error instanceof Error ? error.message : String(error)}`,
      );
    } finally {
      clearTimeout(timeout);
    }
  }
}

export function createHydrationLlmService(
  env: NodeJS.ProcessEnv = process.env,
): HydrationLlmService | null {
  const config = resolveHydrationLlmConfig(env);
  if (config.mode === "deterministic") {
    return null;
  }
  return new OpenAiHydrationLlmService(config);
}

File: merge-tonic-lib/tonic/hydration/llm_service.py
"""Optional provider-backed structured LLM service for hydration flows."""

from __future__ import annotations

import json
from typing import Any, Protocol, TypeVar
from urllib import error, request

from .errors import HydrationRuntimeError
from .llm_config import HydrationLlmConfig, resolve_hydration_llm_config


HydrationLlmRole = str


class HydrationLlmMessage(dict[str, str]):
    pass


T = TypeVar("T")


class HydrationLlmService(Protocol):
    provider_id: str
    model: str

    def generate_json(
        self,
        *,
        messages: list[HydrationLlmMessage],
        schema_name: str,
        schema_description: str,
    ) -> T:
        ...


def _extract_json_object(text: str) -> str:
    trimmed = text.strip()
    if trimmed.startswith("{") and trimmed.endswith("}"):
        return trimmed
    start = trimmed.find("{")
    end = trimmed.rfind("}")
    if start >= 0 and end > start:
        return trimmed[start : end + 1]
    raise HydrationRuntimeError("LLM response did not contain a JSON object.")


class OpenAiHydrationLlmService:
    provider_id = "openai"

    def __init__(self, config: HydrationLlmConfig) -> None:
        if not config.api_key:
            raise HydrationRuntimeError(
                "TONIC_HYDRATION_LLM_MODE=openai requires TONIC_HYDRATION_LLM_API_KEY or OPENAI_API_KEY."
            )
        self._config = config
        self.model = config.model

    def generate_json(
        self,
        *,
        messages: list[HydrationLlmMessage],
        schema_name: str,
        schema_description: str,
    ) -> Any:
        payload = {
            "model": self._config.model,
            "temperature": 0,
            "response_format": {"type": "json_object"},
            "messages": [
                *messages,
                {
                    "role": "system",
                    "content": (
                        f"Return only a JSON object for schema '{schema_name}'. "
                        f"Required shape: {schema_description}"
                    ),
                },
            ],
        }
        req = request.Request(
            f"{self._config.base_url}/chat/completions",
            data=json.dumps(payload).encode("utf-8"),
            headers={
                "Content-Type": "application/json",
                "Authorization": f"Bearer {self._config.api_key}",
            },
            method="POST",
        )
        try:
            with request.urlopen(req, timeout=self._config.timeout_ms / 1000.0) as resp:
                text = resp.read().decode("utf-8", errors="replace")
        except error.HTTPError as exc:
            body = exc.read().decode("utf-8", errors="replace")
            raise HydrationRuntimeError(f"OpenAI HTTP {exc.code}: {body}") from exc
        except OSError as exc:
            raise HydrationRuntimeError(
                f"Failed to generate structured hydration output via {self.provider_id}: {exc}"
            ) from exc

        try:
            parsed = json.loads(text)
            content = (((parsed.get("choices") or [{}])[0] or {}).get("message") or {}).get("content")
            output = (
                content
                if isinstance(content, str)
                else "".join(
                    str(entry.get("text", ""))
                    for entry in content
                    if isinstance(entry, dict)
                )
                if isinstance(content, list)
                else ""
            )
            return json.loads(_extract_json_object(output))
        except (ValueError, TypeError) as exc:
            raise HydrationRuntimeError(
                f"Failed to generate structured hydration output via {self.provider_id}: {exc}"
            ) from exc


def create_hydration_llm_service(env: dict[str, str] | None = None) -> HydrationLlmService | None:
    config = resolve_hydration_llm_config(env)
    if config.mode == "deterministic":
        return None
    return OpenAiHydrationLlmService(config)
